ada_feeding_perception/ada_feeding_perception/face_detection.py

```diff
@@ -0,0 +1,388 @@
+                        read_only=True,
+                    ),
+                ),
+                (
+                    "face_model_base_url",
+                    None,
+                    ParameterDescriptor(
+                        name="face_model_base_url",
+                        type=ParameterType.PARAMETER_STRING,
+                        description=(
+                            "The URL to download the model checkpoint from "
+                            "if it is not already downloaded"
+                        ),
+                        read_only=True,
+                    ),
+                ),
+                (
+                    "landmark_model_name",
+                    None,
+                    ParameterDescriptor(
+                        name="landmark_model_name",
+                        type=ParameterType.PARAMETER_STRING,
+                        description="The name of the face detection model checkpoint to use",
+                        read_only=True,
+                    ),
+                ),
+                (
+                    "landmark_model_base_url",
+                    None,
+                    ParameterDescriptor(
+                        name="landmark_model_base_url",
+                        type=ParameterType.PARAMETER_STRING,
+                        description=(
+                            "The URL to download the model checkpoint from "
+                            "if it is not already downloaded"
+                        ),
+                        read_only=True,
+                    ),
+                ),
+                (
+                    "model_dir",
+                    None,
+                    ParameterDescriptor(
+                        name="model_dir",
+                        type=ParameterType.PARAMETER_STRING,
+                        description=(
+                            "The location of the directory where the model "
+                            "checkpoint is / should be stored"
+                        ),
+                        read_only=True,
+                    ),
+                ),
+            ],
+        )
+
+    def toggle_face_detection_callback(
+        self, request: SetBool.Request, response: SetBool.Response
+    ) -> SetBool.Response:
+        """
+        Callback function for the toggle_face_detection service. Safely toggles
+        the face detection on or off depending on the request.
+        """
+        self.get_logger().info(f"Incoming service request. data: {request.data}")
+        with self.is_on_lock:
+            self.is_on = request.data
+        response.success = request.data
+        return response
+
+    def camera_info_callback(self, msg: CameraInfo):
+        """
+        Callback function for the toggle_face_detection service. Safely toggles
+        the face detection on or off depending on the request.
+        """
+        self.camera_matrix = msg.k
```

Comment:
Since `self.camera_matrix` can be accessed across multiple threads, it should have a lock. It looks like lines 294-298 in `depth_callback` below use it. I'd suggest creating a local copy of the variable at the top of `depth_callback`, like you do with `self.is_on`/`self.is_on_lock`, and then using that local copy (not the instance attribute attribute) throughout the function